affordance_primitives/test/test_task_monitor.cpp

```diff
@@ -0,0 +1,194 @@
+  EXPECT_LE((ros::Time::now() - start).toSec(), 1.05 * test_duration);
+  EXPECT_EQ(result2, affordance_primitives::ExecutionResult::TIME_OUT);
+}
+
+TEST(TaskMonitor, monitor_ft_readings)
+{
+  ros::NodeHandle nh;
+  ros::Publisher ft_pub = nh.advertise<affordance_primitives::WrenchStamped>("ft_topic_name", 1, true);
+
+  const double test_duration = 0.25;
+  const double ft_pub_period = 0.005;  // 200 hz
+  const size_t num_publish = test_duration / ft_pub_period;
+
+  // Set some force limits in parameter
+  affordance_primitives::AffordanceParameter params;
+  params.max_wrench.rot_x = 10;
+  params.max_wrench.rot_y = 10;
+  params.max_wrench.rot_z = 10;
+  params.max_wrench.trans_x = 100;
+  params.max_wrench.trans_y = 100;
+  params.max_wrench.trans_z = 100;
+  params.max_force = 120;
+  params.max_torque = 12;
+
+  // Start publishing FT readings of 0
+  affordance_primitives::WrenchStamped pub_wrench;
+  std::thread t([&ft_pub, &pub_wrench, num_publish, ft_pub_period]() {
+    for (size_t i = 0; i < num_publish; ++i)
+    {
+      ft_pub.publish(pub_wrench);
+      ros::Duration(ft_pub_period).sleep();
+    }
+  });
+
+  // Start the monitor and verify it times out
+  affordance_primitives::TaskMonitor monitor(nh, "ft_topic_name");
+  ASSERT_NO_THROW(monitor.startMonitor(params, test_duration));
+  EXPECT_EQ(monitor.waitForResult(), affordance_primitives::ExecutionResult::TIME_OUT);
+  t.join();
+
+  // Now publish FT readings to violate max force limit
+  t = std::thread([&ft_pub, &pub_wrench, num_publish, ft_pub_period]() {
+    for (size_t i = 0; i < num_publish; ++i)
+    {
+      if (i == num_publish / 3)
+      {
+        pub_wrench.wrench.force.x = 99;
+        pub_wrench.wrench.force.y = 99;
+        pub_wrench.wrench.force.z = 99;
+      }
+      ft_pub.publish(pub_wrench);
+      ros::Duration(ft_pub_period).sleep();
+    }
+  });
+
+  // Start monitor and verify FT limit is violated
+  ASSERT_NO_THROW(monitor.startMonitor(params, test_duration));
+  EXPECT_EQ(monitor.waitForResult(), affordance_primitives::ExecutionResult::FT_VIOLATION);
+  t.join();
+
+  // Now publish to violate individual limits
+  pub_wrench = affordance_primitives::WrenchStamped();
+  t = std::thread([&ft_pub, &pub_wrench, num_publish, ft_pub_period]() {
+    for (size_t i = 0; i < num_publish; ++i)
+    {
+      if (i == num_publish / 3)
+      {
+        pub_wrench.wrench.torque.z = 11;
+      }
+      else
+      {
+        pub_wrench.wrench.torque.z = 9;
+      }
+      ft_pub.publish(pub_wrench);
+      ros::Duration(ft_pub_period).sleep();
+    }
+  });
+
+  // Start monitor and verify FT limit is violated
+  ASSERT_NO_THROW(monitor.startMonitor(params, test_duration));
+  EXPECT_EQ(monitor.waitForResult(), affordance_primitives::ExecutionResult::FT_VIOLATION);
+  t.join();
+}
+
+int main(int argc, char** argv)
+{
+  ros::init(argc, argv, "test_task_monitor");
+  testing::InitGoogleTest(&argc, argv);
+
+  ros::AsyncSpinner spinner(8);
```

Comment:
Just curious, how do you decide the number of cores for your spinner function

Reply:
I copy-pasted from somewhere else :sweat_smile: I am not sure the number actually matters, except that ROS tests should use the multithreaded spinner